patches/minecraft/net/minecraft/item/ItemStack.java.patch

```diff
@@ -230,17 +221,51 @@
 +   public void deserializeNBT(CompoundNBT nbt) {
 +      final ItemStack itemStack = ItemStack.func_199557_a(nbt);
 +      getStack().func_77982_d(itemStack.func_77978_p());
-+      if (itemStack.capNBT != null) deserializeCaps(itemStack.capNBT);
 +   }
 +
 +   /**
-+    * Set up forge's ItemStack additions.
++    * Initialize caps if not already initialized.
 +    */
-+   private void forgeInit() {
-+      if (this.delegate != null) {
++   private void forgeInitCaps() {
++      if (!this.capsInitialized && this.delegate != null) {
```

Comment:
Just a quick remark: I'm not sure what kind of threading guarantee (if any) this is supposed to have. But before it was safe to access caps from multiple threads (AFAICT), now that would be a race condition.

Reply:
What is the usecase for accessing item stacks on multiple threads? In my understanding, most item stack accesses are on the main thread. Is it something you could synchronize in your own code?

Reply:
Good point, in theory someone could be copying a stack on the render thread, while someone is querying `getCapability` on the client thread. Will add a double-checked lock.

Reply:
I don't think any of the gameplay objects in Minecraft are thread-safe.

Reply:
After thinking more about it, the race condition already exists currently. You can mutate the stack's capabilities on one thread and copy it on another thread. That would cause the caps to be modified and serialized at the same time.

Reply:
Should be fine now.